queryable-spec.ts

```diff
@@ -0,0 +1,421 @@
+ * interface, but implementers MAY decide to implement both at the same time.
+ * 
+ * matchExpression() Returns a FilterableSourceResult that contains a quad 
+ * stream that processes all quads matching the quad pattern and the expression.
+ * 
+ * When a Term parameter is defined, and is a NamedNode, Literal or BlankNode, 
+ * it must match each produced quad, according to the Quad.equals semantics. 
+ * When a Term parameter is a Variable, or it is undefined, it acts as a 
+ * wildcard, and can match with any Term.
+ * 
+ * NOTES:
+ * - When matching with graph set to undefined or null it MUST match all the
+ *   graphs (sometimes called the union graph). To match only the default graph
+ *   set graph to a DefaultGraph.
+ * - When an Expression parameter is defined, the complete quad stream is
+ *   filtered according to this expression. When it is undefined, no filter is
+ *   applied.
+ * 
+ * If parameters of type Variable with an equal variable name are in place,
+ * then the corresponding quad components in the resulting quad stream MUST be
+ * equal.
+ * Expression's MAY contain Variable Term's. If their variable names are equal
+ * to Variable's in the given quad pattern, then the Expression MUST be 
+ * instantiated for each variable's binding in the resulting quad stream when
+ * applying the Expression filter.
+ */
+interface FilterableSource {
+  matchExpression(
+    subject?: RDF.Term,
+    predicate?: RDF.Term,
+    obj?: RDF.Term,
+    graph?: RDF.Term,
+    expression?: Expression,
+    opts?: { 
+      limit?: number; 
+      offset?: number; 
+    },
+  ): FilterableResult;
+};
+
+
+/******************************************************************************
+                              QUERYABLE SOURCE
+ *****************************************************************************/
+
+/*
+ * Map-like representation of Bindings as using plain objects could lead
+ * to collisions between variable names and object properties. Support for
+ * immutability is required (but implementations are free to be mutable) which
+ * determines the return value of the set() and delete() methods to be an 
+ * instance of Bindings (potentially a different one).
+ */ 
+interface Bindings {
+  type: 'bindings';
+  has: (key: RDF.Variable) => boolean;
+  get: (key: RDF.Variable) => RDF.Term | undefined;
+  set: (key: RDF.Variable, value: RDF.Term) => Bindings;
+  delete: (key: RDF.Variable) => Bindings;
+  keys: () => Iterator<RDF.Variable>;
+  values: () => Iterator<RDF.Term>;
+  entries: () => Iterator<[RDF.Variable, RDF.Term]>;
+  forEach: (fn: (value: RDF.Term, key: RDF.Variable) => any) => void;
+  size: number;
+}
+
+/*
+ * Bindings objects are created using a dedicated factory, keeping in line
+ * with DataFactory.quad(). This also helps with facilitating support for
+ * immutability. Basic helper methods must also be provided for the most 
+ * common manipulations of bindings objects.
+ */
+interface BindingsFactory {
```

Comment:
Mirroring what RDF/JS had already done with the `DataFactory` interface would be good for long-term compat, I think. One of the possible (and likely, IMHO) use-cases is passing custom factory implementations.

Reply:
I agree with the compat argument in general, but I don't understand the use-case here.

`DataFactory` is for use-cases in which the API consumer wants to construct data to push _into_ the store, or BGP patterns to use in queries. But I don't think this applies to Bindings, which always come _out_ of the store.

When would the API _consumer_ ever construct bindings?

Reply:
This interface will perhaps not be used for API consumers indeed, but it will be used by query engine developers.

Reply:
A consumer of a query engine might itself be a query engine, for example in the case of federation.

Reply:
> This interface will perhaps not be used for API consumers indeed, but it will be used by query engine developers.

Perhaps we need to be clearer what this API is for. I think of this as the query API. As an engine developer you would have lots of considerations beyond this API, of which constructing Bindings is one.

I probably need to stop flogging this... Happy to go with the consensus of course. 😇

Reply:
> Perhaps we need to be clearer what this API is for. I think of this as the query API. As an engine developer you would have lots of considerations beyond this API, of which constructing Bindings is one.

I think the line between engine developers and consumers is much, much blurrier than we're making it out to be. However, even assuming a clear demarcation between the two, there are plenty of "consumer" use-cases that require manipulating bindings at the application level. Unit testing itself requires the ability to create bindings separately from query engines.

Reply:
In the absence of further comments I'm closing this one and keeping bindings creation and manipulation within the spec. As always, happy to reopen and discuss further.